src/ui/layout/mct-tree.vue

```diff
@@ -10,107 +13,452 @@
-                    this.isLoading = false;
-                    this.allTreeItems = children.map(c => {
-                        return {
-                            id: this.openmct.objects.makeKeyString(c.identifier),
-                            object: c,
-                            objectPath: [c],
-                            navigateToParent: '/browse'
-                        };
+        updatevisibleItems() {
+            if (this.updatingView) {
+                return;
+            }
+            this.updatingView = true;
+            requestAnimationFrame(()=> {
+                let start = 0,
+                    end = this.pageThreshold,
+                    allItemsCount = this.focusedItems.length;
+
+                if (allItemsCount < this.pageThreshold) {
+                    end = allItemsCount;
+                } else {
+                    let firstVisible = this.calculateFirstVisibleItem(),
+                        lastVisible = this.calculateLastVisibleItem(),
+                        totalVisible = lastVisible - firstVisible,
+                        numberOffscreen = this.pageThreshold - totalVisible;
+
+                    start = firstVisible - Math.floor(numberOffscreen / 2);
+                    end = lastVisible + Math.ceil(numberOffscreen / 2);
+
+                    if (start < 0) {
+                        start = 0;
+                        end = Math.min(this.pageThreshold, allItemsCount);
+                    } else if (end >= allItemsCount) {
+                        end = allItemsCount;
+                        start = end - this.pageThreshold + 1;
+                    }
+                }
+                this.itemOffset = start;
+                this.visibleItems = this.focusedItems.slice(start, end);
+
+                this.updatingView = false;
+            });
+        },
+        setContainerHeight() {
+            this.$nextTick(() => {
+                let mainTree = this.$refs.mainTree,
+                    mainTreeHeight = mainTree.clientHeight;
+
+                if(mainTreeHeight !== 0) {
+                    this.calculateChildHeight(() => {
+                        let ancestorsHeight = this.calculateAncestorHeight(),
+                            allChildrenHeight = this.calculateChildrenHeight();
+
+                        if(this.activeSearch) {
+                            ancestorsHeight = 0;
+                        }
+                        this.availableContainerHeight = mainTreeHeight - ancestorsHeight;
+
+                        if(allChildrenHeight > this.availableContainerHeight) {
+                            this.setPageThreshold();
+                            this.noScroll = false;
+                        } else {
+                            this.noScroll = true;
+                        }
+                        this.updatevisibleItems();
                     });
+                } else {
+                    window.setTimeout(this.setContainerHeight, RECHECK_DELAY);
```

Comment:
Is there a case where this would need to happen in a timeout?

Reply:
Yeah this is a tricky one, I couldn't have this work reliably without the timeout. Just checked again, if I remove it, the height is all janked up on a fresh load or reload of the page. Open to options on this one...